src/lemmas.k.md

```diff
@@ -64,19 +64,20 @@ rule A -Word B <=Int A => #rangeUInt(256, A -Int B)
 
 ### Bitwise Modulo
 
-The solidity optimizer compiles `block.time % 2**32` into `TIME AND maxUInt32` instead of `TIME
-MOD pow32`. `K` is then unable to reason about the size of the result and so cannot apply further
-simplifications. We therefore rewrite it back to `modInt pow32`.
+The solidity optimizer compiles `block.time % 2**32` to `block.time AND maxUInt32`.
+This lemma ensures the packed storage rules will apply.
 
 ```k
-rule (maxUInt32 &Int X) => (X modInt pow32)
+rule (X &Int Y) <=Int X => true
+  requires X >=Int 0
 ```
 
-Repeated application of `modInt pow32` can be simplified as follows. This lets us clean the storage
+Repeated application of `modInt` can be simplified as follows. This lets us clean the storage
 conditions in a few specs.
 
 ```k
-rule ((X modInt pow32) modInt pow32) => (X modInt pow32)
+rule ((X modInt Y) modInt Y) => (X modInt Y)
+  requires X >=Int 0
```

Comment:
I think that rewriting the lemma above will probably allow us to remove this lemma entirely.

Reply:
Ah interesting. I'm seeing [some failures](https://buildbot.dapp.ci/#/builders/12/builds/591), not sure if they're transient.

Reply:
K is unable to simplify the `chop` away from `chop(((maxUInt32 & TIME) * pow224))`, meaning the packed storage lemmas don't apply, and memory no longer simplifies.

Looks like the rewrite of this lemma does not do what we wanted it to. Perhaps a Z3 lemma will do the trick :thinking:

Reply:
Does K know that `maxUInt32` is 32bits long and pow224 is 224 bits long?

Reply:
I would be very surprised if that was the case

Reply:
maybe this will work?

```
(maxUInt32 &Int X) *Int pow224) <=Int maxUInt256 => true
    requires X >=Int 0
```

Reply:
BTW I'm totally fine with adding a special-case rule like

```
rule chop((maxUInt32 &Int TIME) * pow224) => (maxUInt32 & TIME) * pow224
```

Reply:
Or

```
rule chop((X &Int Y) *Int Z) => (X &Int Y) *Int Z
  requires #rangeUInt(32,  X)
  andBool  #rangeUInt(224, Z)
```

Reply:
fine by me.

Reply:
We agreed elsewhere that that's not necessary.